drivers/gpio/gpio_grgpio2.c

```diff
@@ -0,0 +1,294 @@
+	struct data *data = dev->data;
+	volatile struct grgpio_regs *regs = cfg->regs;
+	uint32_t mask = 1 << pin;
+
+	if (flags & GPIO_SINGLE_ENDED) {
+		return -ENOTSUP;
+	}
+
+	if ((flags & (GPIO_PULL_UP | GPIO_PULL_DOWN)) != 0) {
+		return -ENOTSUP;
+	}
+
+	if (flags & GPIO_OUTPUT) {
+		k_spinlock_key_t key;
+
+		/*
+		 * Register operations are atomic, but do the sequence under
+		 * lock so it serializes.
+		 */
+		key = k_spin_lock(&data->lock);
+		if (flags & GPIO_OUTPUT_INIT_HIGH) {
+			regs->output_or = mask;
+		} else if (flags & GPIO_OUTPUT_INIT_LOW) {
+			regs->output_and = ~mask;
+		}
+		regs->dir_or = mask;
+		k_spin_unlock(&data->lock, key);
+	} else {
+		regs->dir_and = ~mask;
+	}
+
+	return 0;
+}
+
+static int port_get_raw(const struct device *dev, gpio_port_value_t *value)
+{
+	const struct cfg *cfg = dev->config;
+
+	*value = cfg->regs->data;
+	return 0;
+}
+
+static int port_set_masked_raw(const struct device *dev,
+					  gpio_port_pins_t mask,
+					  gpio_port_value_t value)
+{
+	const struct cfg *cfg = dev->config;
+	struct data *data = dev->data;
+	volatile struct grgpio_regs *regs = cfg->regs;
+	uint32_t port_val;
+	k_spinlock_key_t key;
+
+	value &= mask;
+	key = k_spin_lock(&data->lock);
+	port_val = (regs->output & ~mask) | value;
+	regs->output = port_val;
+	k_spin_unlock(&data->lock, key);
+
+	return 0;
+}
+
+static int port_set_bits_raw(const struct device *dev, gpio_port_pins_t pins)
+{
+	const struct cfg *cfg = dev->config;
+	volatile struct grgpio_regs *regs = cfg->regs;
+
+	regs->output_or = pins;
+	return 0;
+}
+
+static int port_clear_bits_raw(const struct device *dev, gpio_port_pins_t pins)
+{
+	const struct cfg *cfg = dev->config;
+	volatile struct grgpio_regs *regs = cfg->regs;
+
+	regs->output_and = ~pins;
+	return 0;
+}
+
+static int port_toggle_bits(const struct device *dev, gpio_port_pins_t pins)
+{
+	const struct cfg *cfg = dev->config;
+	volatile struct grgpio_regs *regs = cfg->regs;
+
+	regs->output_xor = pins;
+	return 0;
+}
+
+static uint32_t get_pending_int(const struct device *dev)
+{
+	const struct cfg *cfg = dev->config;
+	volatile struct grgpio_regs *regs = cfg->regs;
+
+	return regs->iflag;
+}
+
+static int pin_interrupt_configure(const struct device *dev,
+					      gpio_pin_t pin,
+					      enum gpio_int_mode mode,
+					      enum gpio_int_trig trig)
+{
+	const struct cfg *cfg = dev->config;
+	struct data *data = dev->data;
+	volatile struct grgpio_regs *regs = cfg->regs;
+	int ret = 0;
+	const uint32_t mask = 1 << pin;
+	uint32_t polmask;
+	k_spinlock_key_t key;
+
+	if ((mask & data->imask) == 0) {
+		/* This pin can not generate interrupt */
+		return -ENOTSUP;
+	}
+	if (mode != GPIO_INT_MODE_DISABLED) {
+		if (trig == GPIO_INT_TRIG_LOW) {
+			polmask = 0;
+		} else if (trig == GPIO_INT_TRIG_HIGH) {
+			polmask = mask;
+		} else {
+			return -ENOTSUP;
+		}
+	}
+	key = k_spin_lock(&data->lock);
+	if (mode == GPIO_INT_MODE_DISABLED) {
+		regs->imask_and = ~mask;
```

Comment:
If we cleared here respective bit in `regs->iflag` register would that solve the issue with level interrupts?

Reply:
Actually, it would be more appropriate to clear `regs->iflag` shortly before enabling the interrupt and not after disabling it.

Reply:
Agreed. The respective bit in `iflag` should be cleared shortly before enabling the interrupt. It has been added in the latest push. The flag gets cleared also when disabling interrupt for a pin, and that solved the issue with level interrupts in the test case. Thanks!

All subtests in `gpio_api_1pin_test` now pass.